File: frontend/src/services/audit-service.ts
import apiClient from "@/lib/api-client";
import type {
  AuditLogEntry,
  AuditLogFilters,
  PaginatedAuditLogs,
} from "@/types/audit";

export const auditService = {
  async getAuditLogs(
    page = 1,
    perPage = 50,
    filters?: AuditLogFilters
  ): Promise<PaginatedAuditLogs> {
    const params = new URLSearchParams({
      page: String(page),
      per_page: String(perPage),
    });
    if (filters?.user_id) params.set("user_id", filters.user_id);
    if (filters?.action) params.set("action", filters.action);
    if (filters?.entity_type) params.set("entity_type", filters.entity_type);
    if (filters?.entity_id) params.set("entity_id", filters.entity_id);
    if (filters?.date_from) params.set("date_from", filters.date_from);
    if (filters?.date_to) params.set("date_to", filters.date_to);

    const response = await apiClient.get<PaginatedAuditLogs>(
      `/audit-logs?${params.toString()}`
    );
    return response.data;
  },

  async getAuditLog(logId: string): Promise<AuditLogEntry> {
    const response = await apiClient.get<AuditLogEntry>(
      `/audit-logs/${logId}`
    );
    return response.data;
  },
};

File: backend/app/services/audit_service.py
"""
Audit logging service — append-only event log for all business operations.

Usage from any service:
    from app.services import audit_service

    audit_service.log_action(
        db, company_id, "created", "user", user.id,
        user_id=current_user.id,
        changes={"email": "new@example.com"},
    )
"""

import json
from datetime import datetime

from sqlalchemy.orm import Session

from app.models.audit_log import AuditLog


def log_action(
    db: Session,
    company_id: str,
    action: str,
    entity_type: str,
    entity_id: str | None = None,
    user_id: str | None = None,
    changes: dict | None = None,
    ip_address: str | None = None,
    user_agent: str | None = None,
) -> AuditLog:
    """
    Record an audit event. Uses db.flush() (not commit) so the audit entry
    is committed atomically with the caller's business operation.
    """
    entry = AuditLog(
        company_id=company_id,
        user_id=user_id,
        action=action,
        entity_type=entity_type,
        entity_id=entity_id,
        changes=json.dumps(changes) if changes else None,
        ip_address=ip_address,
        user_agent=user_agent,
    )
    db.add(entry)
    db.flush()
    return entry


def compute_changes(old: dict, new: dict) -> dict | None:
    """
    Compare old and new values, return a dict of changed fields.
    Format: {"field_name": {"old": old_value, "new": new_value}}
    Returns None if no changes.
    """
    changes = {}
    for key in new:
        if key in old and old[key] != new[key]:
            changes[key] = {"old": old[key], "new": new[key]}
    return changes if changes else None


def get_audit_logs(
    db: Session,
    company_id: str,
    page: int = 1,
    per_page: int = 50,
    user_id: str | None = None,
    action: str | None = None,
    entity_type: str | None = None,
    entity_id: str | None = None,
    date_from: datetime | None = None,
    date_to: datetime | None = None,
) -> dict:
    """Query audit logs with pagination and filtering."""
    query = db.query(AuditLog).filter(AuditLog.company_id == company_id)

    if user_id:
        query = query.filter(AuditLog.user_id == user_id)
    if action:
        query = query.filter(AuditLog.action == action)
    if entity_type:
        query = query.filter(AuditLog.entity_type == entity_type)
    if entity_id:
        query = query.filter(AuditLog.entity_id == entity_id)
    if date_from:
        query = query.filter(AuditLog.created_at >= date_from)
    if date_to:
        query = query.filter(AuditLog.created_at <= date_to)

    total = query.count()
    items = (
        query.order_by(AuditLog.created_at.desc())
        .offset((page - 1) * per_page)
        .limit(per_page)
        .all()
    )
    return {"items": items, "total": total, "page": page, "per_page": per_page}


def get_audit_log(db: Session, log_id: str, company_id: str) -> AuditLog | None:
    """Fetch a single audit log entry."""
    return (
        db.query(AuditLog)
        .filter(AuditLog.id == log_id, AuditLog.company_id == company_id)
        .first()
    )
